README
-----Ruby/SerialPort-----

-- !Please Note! --

This is an experimental version of the official ruby-serialport, found at http://rubygems.org/gems/serialport
Use with care! Added listDevices to find available ports on the system.



-- Description --

Ruby/SerialPort is a Ruby library that provides a class for using
RS-232 serial ports.  This class also contains low-level functions to
check and set the current state of the signals on the line.

The native Windows version of this library supports Microsoft's Visual C++, Borland's C++, and MinGW compilers.

-- Installation --

This gem is hosted at Gemcutter. Before you can install this, or any other Gemcutter gem you must
add it as a gem source:
  gem sources -a http://gemcutter.org

Then you can install the gem as normal:
  sudo gem install serialport

-- Testing --

* test/miniterm.rb

Ruby's copy of miniterm.c !


-- API --

  **** Class SerialPort, Parent IO ****

    ** Class constants **

       VERSION -> aString (this release is "0.6")
       NONE, HARD, SOFT, SPACE, MARK, EVEN, ODD -> anInteger

    ** Class methods **
	  * listDevices() -> an Array of Hashes with ports found (OS X only, for now)
      * new(port_num [, modem_parameters]) -> aSerialPort
      * open(port_num [, modem_parameters]) -> aSerialPort
      * open(port_num [, modem_parameters]) {|aSerialPort| block} ->
                value of the block

        port_num -> anInteger: port number, 0 for first port which is
                      "/dev/ttyS0" on GNU/Linux and "COM1" on Windows,
                or  aString: file name of the device (example: "/dev/ttyS2")

        Optional modem_parameters:

        baudrate -> anInteger: from 50 to 256000, depends on platform.

        databits -> anInteger: from 5 to 8 (4 is allowed on Windows)

        stopbits -> anInteger: 1 or 2 (1.5 is not supported)

        parity -> anInteger: SerialPort::NONE, SerialPort::EVEN,
                  SerialPort::ODD, SerialPort::MARK, SerialPort::SPACE
                  (MARK and SPACE are not supported on Posix)

        Raise an argError on bad argument.

        SerialPort::new and SerialPort::open without a block return an
        instance of SerialPort.  SerialPort::open with a block passes
        a SerialPort to the block and closes it when the block exits
        (like File::open).


    ** Instance methods **

      * modem_params() -> aHash
      * modem_params=(aHash) -> aHash
      * get_modem_params() -> aHash
      * set_modem_params(aHash) -> aHash
      * set_modem_params(baudrate [, databits [, stopbits [, parity]]])

        Get and set the modem parameters.  Hash keys are "baud", "data_bits",
        "stop_bits", and "parity" (see above).

        Parameters not present in the hash or set to nil remain unchanged.
        Default parameter values for the set_modem_params method are:
        databits = 8, stopbits = 1, parity = (databits == 8 ?
        SerialPort::NONE : SerialPort::EVEN).

      * baud() -> anInteger
      * baud=(anInteger) -> anInteger
      * data_bits() -> 4, 5, 6, 7, or 8
      * data_bits=(anInteger) -> anInteger
      * stop_bits() -> 1 or 2
      * stop_bits=(anInteger) -> anInteger
      * parity() -> anInteger: SerialPort::NONE, SerialPort::EVEN,
          SerialPort::ODD, SerialPort::MARK, or SerialPort::SPACE
      * parity=(anInteger) -> anInteger

        Get and set the corresponding modem parameter.

      * flow_control() -> anInteger
      * flow_control=(anInteger) -> anInteger

        Get and set the flow control: SerialPort::NONE, SerialPort::HARD,
        SerialPort::SOFT, or (SerialPort::HARD | SerialPort::SOFT).

        Note: SerialPort::HARD mode is not supported on all platforms.
        SerialPort::HARD uses RTS/CTS handshaking; DSR/DTR is not
        supported.

      * read_timeout() -> anInteger
      * read_timeout=(anInteger) -> anInteger
      * write_timeout() -> anInteger
      * write_timeout=(anInteger) -> anInteger

        Get and set timeout values (in milliseconds) for reading and writing.
        A negative read timeout will return all the available data without
        waiting, a zero read timeout will not return until at least one
        byte is available, and a positive read timeout returns when the
        requested number of bytes is available or the interval between the
        arrival of two bytes exceeds the timeout value.

        Note: Read timeouts don't mix well with multi-threading.

        Note: Under Posix, write timeouts are not implemented.

      * break(time) -> nil

        Send a break for the given time.  

        time -> anInteger: tenths-of-a-second for the break.
        Note: Under Posix, this value is very approximate.

      * signals() -> aHash

        Return a hash with the state of each line status bit.  Keys are
        "rts", "dtr", "cts", "dsr", "dcd", and "ri".

        Note: Under Windows, the rts and dtr values are not included.

      * rts()
      * dtr()
      * cts()
      * dsr()
      * dcd()
      * ri() -> 0 or 1

      * rts=(0 or 1)
      * dtr=(0 or 1) -> 0 or 1

        Get and set the corresponding line status bit.

        Note: Under Windows, rts() and dtr() are not implemented.

-- License --

GPL

Guillaume Pierronnet <moumar@netcourrier.com>
Alan Stern <stern@rowland.harvard.edu>
Tobin Richard <tobin.richard@gmail.com>
Hector Parra <hector@hectorparra.com>
Ryan C. Payne <rpayne-oss@bullittsystems.com>
Jens Alexander Ewald <jens@ififelse.net>



CHANGELOG
1.0.5 => 03/05/2011:	[NEW] [OSX] Listing of Devices is supported as ClassMethod (SerialPort.listDevices), returning an array of info hashes (still experimental!)

1.0.4 => 07/17/2010:	[FIXED] [Windows] No longer restricted to specific bitrates
						[FIXED] [Windows] Removed potentially insecure sprintf() use
						[FIXED] [Windows] Workaround for rb_sys_fail not checking GetLastError()
						
1.0.3 => 04/08/2010:	[FIXED] [Windows] Could not specify 10 or "COM10" or higher 
						[FIXED] [Windows] Warning passing INT instead of LONG to GetCommModemStatus

1.0.2 => 04/06/2010:	[FIXED] Passing a block into open did not properly handle a return from within the block

1.0.1 => 01/20/2010:   	[FIXED] Conditional RB_SERIAL_EXPORT needed for Visual Studio bonked GCC

1.0.0 => 01/12/2010:    [FIXED] x86_64 segmentation faults
						[NEW] Windows Ruby 1.9 support

0.7.4 => 10/12/2009:    [NEW] Conditional 1.8.6 & 1.9 support (POSIX only).
						[NEW] MinGW support.

0.7.3 => 10/09/2009:    [NEW] POSIX Ruby 1.9 support

0.7.2 => 02/09/2008:    Fix several Windows and POSIX compilation problems.
                        Make GCC compile the code without any warnings.
                        Make RDoc not be so talkative during gem installation.

0.7.1 => 31/08/2008:    Change to gemspec creation and fix posix compilation.

0.7.0 => 03/07/2008:    Major Code Cleanup

0.6.1 => 25/03/2003:    Minor changes

0.6 => 12/02/2003:      Windows support
                        Get/set modem parameters
                        Read and write timeouts
                        Open method

0.5 => 25/10/2002:      Cygwin support

0.4 => 19/09/2002:      Added more serial ports (total of 8 ports)

0.3 => 15/03/2002:      Damn, another bug found

0.2 => 14/03/2002:      A bug fixed (read() was not blocking)

0.1 => 14/03/2002:      First release


test/miniterm.rb
#!/usr/bin/ruby
require "rubygems"
require "serialport"


if ARGV.size < 4
  STDERR.print <<EOF
  Usage: ruby #{$0} num_port bps nbits stopb
EOF
  exit(1)
end

sp = SerialPort.new(ARGV[0], ARGV[1].to_i, ARGV[2].to_i, ARGV[3].to_i, SerialPort::NONE)

open("/dev/tty", "r+") { |tty|
  tty.sync = true
  Thread.new {
    while true do
      tty.printf("%c", sp.getc)
    end
  }
  while (l = tty.gets) do
    sp.write(l.sub("\n", "\r"))
  end
}

sp.close


test/set_readtimeout.rb
#!/usr/bin/ruby
require "rubygems"
require "serialport"

sp = SerialPort.new(SerialPort.listDevices.first["port"], "38400".to_i)
sp.read_timeout = 100
sp.close


ext/native/serialport.h
/* Ruby/SerialPort
 * Guillaume Pierronnet <moumar@netcourrier.com>
 * Alan Stern <stern@rowland.harvard.edu>
 * Daniel E. Shipton <dshipton@redshiptechnologies.com>
 * Tobin Richard <tobin.richard@gmail.com>
 * Ryan C. Payne <rpayne-oss@bullittsystems.com>
 *
 * This code is hereby licensed for public consumption under either the
 * GNU GPL v2 or greater.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * For documentation on serial programming, see the excellent:
 * "Serial Programming Guide for POSIX Operating Systems"
 * written Michael R. Sweet.
 * http://www.easysw.com/~mike/serial/
 */

#ifndef _RUBY_SERIAL_PORT_H_
#define _RUBY_SERIAL_PORT_H_

#define RUBY_SERIAL_PORT_VERSION   "1.0.4"

#include <ruby.h>    /* ruby inclusion */
#ifdef HAVE_RUBY_IO_H      /* ruby io inclusion */
   #include <ruby/io.h>
#else
   #include <rubyio.h>
#endif

struct modem_params
{
   int data_rate;
   int data_bits;
   int stop_bits;
   int parity;
};

struct line_signals
{
   int rts;
   int dtr;
   int cts;
   int dsr;
   int dcd;
   int ri;
};

#define NONE   0
#define HARD   1
#define SOFT   2

#if defined(OS_MSWIN) || defined(OS_BCCWIN) || defined(OS_MINGW)
   #define SPACE  SPACEPARITY
   #define MARK   MARKPARITY
   #define EVEN   EVENPARITY
   #define ODD    ODDPARITY 

   #ifndef RB_SERIAL_EXPORT
	 #ifndef HAVE_RUBY_IO_H
     	#define RB_SERIAL_EXPORT __declspec(dllexport)
	 #else
		#define RB_SERIAL_EXPORT
	 #endif
   #endif

#else
   #define SPACE  0
   #define MARK   0
   #define EVEN   1
   #define ODD    2
   
   #define RB_SERIAL_EXPORT
#endif

extern VALUE sBaud, sDataBits, sStopBits, sParity; /* strings */
extern VALUE sRts, sDtr, sCts, sDsr, sDcd, sRi;

/* Implementation specific functions. */
VALUE RB_SERIAL_EXPORT sp_create_impl(VALUE class, VALUE _port);
VALUE RB_SERIAL_EXPORT sp_set_modem_params_impl(int argc, VALUE *argv, VALUE self);
void RB_SERIAL_EXPORT get_modem_params_impl(VALUE self, struct modem_params *mp);
VALUE RB_SERIAL_EXPORT sp_set_flow_control_impl(VALUE self, VALUE val);
VALUE RB_SERIAL_EXPORT sp_get_flow_control_impl(VALUE self);
VALUE RB_SERIAL_EXPORT sp_set_read_timeout_impl(VALUE self, VALUE val);
VALUE RB_SERIAL_EXPORT sp_get_read_timeout_impl(VALUE self);
VALUE RB_SERIAL_EXPORT sp_set_write_timeout_impl(VALUE self, VALUE val);
VALUE RB_SERIAL_EXPORT sp_get_write_timeout_impl(VALUE self);
VALUE RB_SERIAL_EXPORT sp_break_impl(VALUE self, VALUE time);
void RB_SERIAL_EXPORT get_line_signals_helper_impl(VALUE obj, struct line_signals *ls);
VALUE RB_SERIAL_EXPORT set_signal_impl(VALUE obj, VALUE val, int sig);
VALUE RB_SERIAL_EXPORT sp_set_rts_impl(VALUE self, VALUE val);
VALUE RB_SERIAL_EXPORT sp_set_dtr_impl(VALUE self, VALUE val);
VALUE RB_SERIAL_EXPORT sp_get_rts_impl(VALUE self);
VALUE RB_SERIAL_EXPORT sp_get_dtr_impl(VALUE self);

#endif


ext/native/serialport.c
/* Ruby/SerialPort
 * Guillaume Pierronnet <moumar@netcourrier.com>
 * Alan Stern <stern@rowland.harvard.edu>
 * Daniel E. Shipton <dshipton@redshiptechnologies.com>
 * Jonas Bähr <jonas.baehr@fs.ei.tum.de>
 * Ryan C. Payne <rpayne-oss@bullittsystems.com>
 * Jens Alexander Ewald <jens@ififelse.net>
 *
 * This code is hereby licensed for public consumption under either the
 * GNU GPL v2 or greater.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * For documentation on serial programming, see the excellent:
 * "Serial Programming Guide for POSIX Operating Systems"
 * written Michael R. Sweet.
 * http://www.easysw.com/~mike/serial/
 */

#include "serialport.h"

VALUE cSerialPort; /* serial port class */

VALUE sBaud, sDataBits, sStopBits, sParity; /* strings */
VALUE sRts, sDtr, sCts, sDsr, sDcd, sRi;

/*
 * :nodoc: This method is private and will be called by SerialPort#new or SerialPort#open.
 */
static VALUE sp_create(class, _port)
   VALUE class, _port;
{
   return sp_create_impl(class, _port);
}


/*
 *  List all available devices (ClassMethod, OS X only for now)
 */
static VALUE sp_listDevices(class)
VALUE class;
{
    return sp_listDevices_impl(class);
}


/*
 * Configure the serial port. You can pass a hash or multiple values
 * as separate arguments. Invalid or unsupported values will raise
 * an ArgumentError.
 *
 * When using a hash the following keys are recognized:
 * ["baud"] Integer from 50 to 256000, depends on platform
 * ["data_bits"] Integer from 5 to 8 (4 is allowed on Windows too)
 * ["stop_bits"] An integer, only allowed values are 1 or 2 (1.5 is not supported)
 * ["parity"] One of the constants NONE, EVEN or ODD (Windows allows also MARK and SPACE)
 *
 * When using separate arguments, they are interpreted as:
 * (baud, data_bits = 8, stop_bits = 1, parity = (previous_databits==8 ? NONE : EVEN))
 *
 * Nota: A baudrate of nil will keep the old value. The default parity depends on the
 * number of databits configured before this function call.
 */
static VALUE sp_set_modem_params(argc, argv, self)
   int argc;
   VALUE *argv, self;
{
   return sp_set_modem_params_impl(argc, argv, self);
}

/*
 * Send a break for the given time.
 *
 * <tt>time</tt> is an integer of tenths-of-a-second for the break.
 *
 * Note: Under Posix, this value is very approximate.
 */
static VALUE sp_break(self, time)
   VALUE self, time;
{
   return sp_break_impl(self, time);
}

/*
 * Get the state (0 or 1) of the DTR line (not available on Windows)
 */
static VALUE sp_get_dtr(self)
   VALUE self;
{
   return sp_get_dtr_impl(self);
}

/*
 * Get the flow control. The result is either NONE, HARD, SOFT or (HARD | SOFT)
 */
static VALUE sp_get_flow_control(self)
   VALUE self;
{
   return sp_get_flow_control_impl(self);
}

/*
 * Get the timeout value (in milliseconds) for reading.
 * See SerialPort#set_read_timeout for details.
 */
static VALUE sp_get_read_timeout(self)
   VALUE self;
{
   return sp_get_read_timeout_impl(self);
}

/*
 * Get the state (0 or 1) of the RTS line (not available on Windows)
 */
static VALUE sp_get_rts(self)
   VALUE self;
{
   return sp_get_rts_impl(self);
}

/*
 * Get the write timeout (in milliseconds)
 *
 * Note: Under Posix, write timeouts are not implemented.
 */
static VALUE sp_get_write_timeout(self)
   VALUE self;
{
   return sp_get_write_timeout_impl(self);
}

/*
 * Set the state (0 or 1) of the DTR line
 */
static VALUE sp_set_dtr(self, val)
   VALUE self, val;
{
   return sp_set_dtr_impl(self, val);
}

/*
 * Set the flow control to either NONE, HARD, SOFT or (HARD | SOFT)
 *
 *  Note: SerialPort::HARD mode is not supported on all platforms.
 *  SerialPort::HARD uses RTS/CTS handshaking; DSR/DTR is not
 *  supported.
 */
static VALUE sp_set_flow_control(self, val)
   VALUE self, val;
{
   return sp_set_flow_control_impl(self, val);
}

/*
 * Set the timeout value (in milliseconds) for reading.
 * A negative read timeout will return all the available data without
 * waiting, a zero read timeout will not return until at least one
 * byte is available, and a positive read timeout returns when the
 * requested number of bytes is available or the interval between the
 * arrival of two bytes exceeds the timeout value.
 *
 * Note: Read timeouts don't mix well with multi-threading.
 */
static VALUE sp_set_read_timeout(self, val)
   VALUE self, val;
{
   return sp_set_read_timeout_impl(self, val);
}

/*
 * Set the state (0 or 1) of the RTS line
 */
static VALUE sp_set_rts(self, val)
   VALUE self, val;
{
   return sp_set_rts_impl(self, val);
}

/*
 * Set a write timeout (in milliseconds)
 *
 * Note: Under Posix, write timeouts are not implemented.
 */
static VALUE sp_set_write_timeout(self, val)
   VALUE self, val;
{
   return sp_set_write_timeout_impl(self, val);
}

/*
 */
static void get_modem_params(self, mp)
   VALUE self;
   struct modem_params *mp;
{
   get_modem_params_impl(self, mp);
}

/*
 * Set the baud rate, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_data_rate(self, data_rate)
   VALUE self, data_rate;
{
   VALUE argv[4];

   argv[0] = data_rate;
   argv[1] = argv[2] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return data_rate;
}

/*
 * Set the data bits, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_data_bits(self, data_bits)
   VALUE self, data_bits;
{
   VALUE argv[4];

   argv[1] = data_bits;
   argv[0] = argv[2] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return data_bits;
}

/*
 * Set the stop bits, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_stop_bits(self, stop_bits)
   VALUE self, stop_bits;
{
   VALUE argv[4];

   argv[2] = stop_bits;
   argv[0] = argv[1] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return stop_bits;
}

/*
 * Set the parity, see SerialPort#set_modem_params for details.
 */
static VALUE sp_set_parity(self, parity)
   VALUE self, parity;
{
   VALUE argv[4];

   argv[3] = parity;
   argv[0] = argv[1] = argv[2] = Qnil;
   sp_set_modem_params(4, argv, self);

   return parity;
}

/*
 * Get the current baud rate, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_data_rate(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.data_rate);
}

/*
 * Get the current data bits, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_data_bits(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.data_bits);
}

/*
 * Get the current stop bits, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_stop_bits(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.stop_bits);
}

/*
 * Get the current parity, see SerialPort#get_modem_params for details.
 */
static VALUE sp_get_parity(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.parity);
}

/*
 * Get the configure of the serial port.
 *
 * Returned is a hash with the following keys:
 * ["baud"] Integer with the baud rate
 * ["data_bits"] Integer from 5 to 8 (4 is possible on Windows too)
 * ["stop_bits"] Integer, 1 or 2 (1.5 is not supported)
 * ["parity"] One of the constants NONE, EVEN or ODD (on Windows may also MARK or SPACE)
 */
static VALUE sp_get_modem_params(self)
   VALUE self;
{
   struct modem_params mp;
   VALUE hash;

   get_modem_params(self, &mp);

   hash = rb_hash_new();

   rb_hash_aset(hash, sBaud, INT2FIX(mp.data_rate));
   rb_hash_aset(hash, sDataBits, INT2FIX(mp.data_bits));
   rb_hash_aset(hash, sStopBits, INT2FIX(mp.stop_bits));
   rb_hash_aset(hash, sParity, INT2FIX(mp.parity));

   return hash;
}

void get_line_signals_helper(obj, ls)
   VALUE obj;
   struct line_signals *ls;
{
   get_line_signals_helper_impl(obj, ls);
}

/*
 * Get the state (0 or 1) of the CTS line
 */
static VALUE sp_get_cts(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.cts);
}

/*
 * Get the state (0 or 1) of the DSR line
 */
static VALUE sp_get_dsr(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.dsr);
}

/*
 * Get the state (0 or 1) of the DCD line
 */
static VALUE sp_get_dcd(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.dcd);
}

/*
 * Get the state (0 or 1) of the RI line
 */
static VALUE sp_get_ri(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.ri);
}

/*
 * Return a hash with the state of each line status bit.  Keys are
 * "rts", "dtr", "cts", "dsr", "dcd", and "ri".
 *
 * Note: Under Windows, the rts and dtr values are not included.
 */
static VALUE sp_signals(self)
   VALUE self;
{
   struct line_signals ls;
   VALUE hash;

   get_line_signals_helper(self, &ls);

   hash = rb_hash_new();

#if !(defined(OS_MSWIN) || defined(OS_BCCWIN) || defined(OS_MINGW))
   rb_hash_aset(hash, sRts, INT2FIX(ls.rts));
   rb_hash_aset(hash, sDtr, INT2FIX(ls.dtr));
#endif
   rb_hash_aset(hash, sCts, INT2FIX(ls.cts));
   rb_hash_aset(hash, sDsr, INT2FIX(ls.dsr));
   rb_hash_aset(hash, sDcd, INT2FIX(ls.dcd));
   rb_hash_aset(hash, sRi, INT2FIX(ls.ri));

   return hash;
}

/*
 * This class is used for communication over a serial port.
 * In addition to the methods here, you can use everything
 * Ruby's IO-class provides (read, write, getc, readlines, ...)
 */
void Init_serialport()
{
   sBaud = rb_str_new2("baud");
   sDataBits = rb_str_new2("data_bits");
   sStopBits = rb_str_new2("stop_bits");
   sParity = rb_str_new2("parity");
   sRts = rb_str_new2("rts");
   sDtr = rb_str_new2("dtr");
   sCts = rb_str_new2("cts");
   sDsr = rb_str_new2("dsr");
   sDcd = rb_str_new2("dcd");
   sRi = rb_str_new2("ri");

   rb_gc_register_address(&sBaud);
   rb_gc_register_address(&sDataBits);
   rb_gc_register_address(&sStopBits);
   rb_gc_register_address(&sParity);
   rb_gc_register_address(&sRts);
   rb_gc_register_address(&sDtr);
   rb_gc_register_address(&sCts);
   rb_gc_register_address(&sDsr);
   rb_gc_register_address(&sDcd);
   rb_gc_register_address(&sRi);

   cSerialPort = rb_define_class("SerialPort", rb_cIO);
   rb_define_singleton_method(cSerialPort, "create", sp_create, 1);
   rb_define_singleton_method(cSerialPort, "listDevices", sp_listDevices, 0);

   rb_define_method(cSerialPort, "get_modem_params", sp_get_modem_params, 0);
   rb_define_method(cSerialPort, "set_modem_params", sp_set_modem_params, -1);
   rb_define_method(cSerialPort, "modem_params", sp_get_modem_params, 0);
   rb_define_method(cSerialPort, "modem_params=", sp_set_modem_params, -1);
   rb_define_method(cSerialPort, "baud", sp_get_data_rate, 0);
   rb_define_method(cSerialPort, "baud=", sp_set_data_rate, 1);
   rb_define_method(cSerialPort, "data_bits", sp_get_data_bits, 0);
   rb_define_method(cSerialPort, "data_bits=", sp_set_data_bits, 1);
   rb_define_method(cSerialPort, "stop_bits", sp_get_stop_bits, 0);
   rb_define_method(cSerialPort, "stop_bits=", sp_set_stop_bits, 1);
   rb_define_method(cSerialPort, "parity", sp_get_parity, 0);
   rb_define_method(cSerialPort, "parity=", sp_set_parity, 1);

   rb_define_method(cSerialPort, "flow_control=", sp_set_flow_control, 1);
   rb_define_method(cSerialPort, "flow_control", sp_get_flow_control, 0);

   rb_define_method(cSerialPort, "read_timeout", sp_get_read_timeout, 0);
   rb_define_method(cSerialPort, "read_timeout=", sp_set_read_timeout, 1);
   rb_define_method(cSerialPort, "write_timeout", sp_get_write_timeout, 0);
   rb_define_method(cSerialPort, "write_timeout=", sp_set_write_timeout, 1);

   rb_define_method(cSerialPort, "break", sp_break, 1);

   rb_define_method(cSerialPort, "signals", sp_signals, 0);
   rb_define_method(cSerialPort, "get_signals", sp_signals, 0);
   rb_define_method(cSerialPort, "rts",  sp_get_rts, 0);
   rb_define_method(cSerialPort, "rts=", sp_set_rts, 1);
   rb_define_method(cSerialPort, "dtr",  sp_get_dtr, 0);
   rb_define_method(cSerialPort, "dtr=", sp_set_dtr, 1);
   rb_define_method(cSerialPort, "cts",  sp_get_cts, 0);
   rb_define_method(cSerialPort, "dsr",  sp_get_dsr, 0);
   rb_define_method(cSerialPort, "dcd",  sp_get_dcd, 0);
   rb_define_method(cSerialPort, "ri",   sp_get_ri, 0);

   rb_define_const(cSerialPort, "NONE", INT2FIX(NONE));
   rb_define_const(cSerialPort, "HARD", INT2FIX(HARD));
   rb_define_const(cSerialPort, "SOFT", INT2FIX(SOFT));

   rb_define_const(cSerialPort, "SPACE", INT2FIX(SPACE));
   rb_define_const(cSerialPort, "MARK", INT2FIX(MARK));
   rb_define_const(cSerialPort, "EVEN", INT2FIX(EVEN));
   rb_define_const(cSerialPort, "ODD", INT2FIX(ODD));

   /* the package's version as a string "X.Y.Z", beeing major, minor and patch level */
   rb_define_const(cSerialPort, "VERSION", rb_str_new2(RUBY_SERIAL_PORT_VERSION));
}
